File: dsa_ninjas_java/Stacks & Queues/StockSpanProblem.java
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

public class StockSpanProblem {
    public static List<Integer> findStockSpan(Integer[] input) {
        class StackData {
            Integer element;
            Integer index;

            StackData(Integer element, Integer index) {
                this.element = element;
                this.index = index;
            }
        }

        Stack<StackData> stack = new Stack<>();
        List<Integer> result = new LinkedList<>();
        for (Integer currentIndex = 0; currentIndex < input.length; currentIndex++) {
            while (!stack.isEmpty() && stack.peek().element < input[currentIndex]) {
                stack.pop();
            }
            if (stack.isEmpty())
                result.add(currentIndex + 1);
            else
                result.add(currentIndex - stack.peek().index);
            stack.push(new StackData(input[currentIndex], currentIndex));
        }

        return result;
    }

    public static void main(String[] args) {
        System.out.println(findStockSpan(new Integer[] { 60, 70, 80, 100, 90, 75, 80, 120 }));
        System.out.println(findStockSpan(new Integer[] { 10, 10, 10, 10 }));
    }
}

File: dsa_ninjas_java/Stacks & Queues/stock_span_problem.py
def stockSpan(price_list):
    stack, result = [], []
    for current_index in range(len(price_list)):
        while stack and stack[-1][0] < price_list[current_index]:
            stack.pop()
        if stack:
            result.append(current_index - stack[-1][1])
        else:
            result.append(current_index+1)
        stack.append([price_list[current_index], current_index])
    return result

print(stockSpan([60, 70, 80, 100, 90, 75, 80, 120]))
print(stockSpan([10, 10, 10, 10]))